x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java

```diff
@@ -202,16 +191,64 @@ private void includeAndExcludeFields(Set<String> fields) {
         }
     }
 
-    private void checkRequiredFieldsArePresent(Set<String> fields) {
-        List<String> missingFields = config.getAnalysis().getRequiredFields()
-            .stream()
-            .filter(f -> fields.contains(f) == false)
-            .collect(Collectors.toList());
-        if (missingFields.isEmpty() == false) {
-            throw ExceptionsHelper.badRequestException("required fields {} are missing", missingFields);
+    private void checkNoIgnoredFields(Set<String> fields) {
+        Optional<String> ignoreField = IGNORE_FIELDS.stream().filter(fields::contains).findFirst();
+        if (ignoreField.isPresent()) {
+            throw ExceptionsHelper.badRequestException("field [{}] cannot be analyzed", ignoreField.get());
         }
     }
 
+    private void checkFieldsHaveCompatibleTypes(Set<String> fields) {
+        for (String field : fields) {
+            Map<String, FieldCapabilities> fieldCaps = fieldCapabilitiesResponse.getField(field);
+            if (fieldCaps == null) {
+                throw ExceptionsHelper.badRequestException("no mappings could be found for field [{}]", field);
+            }
+
+            // Check all field has compatible type
+            Set<String> asSet = new HashSet<>(Collections.singleton(field));
+            removeFieldsWithIncompatibleTypes(asSet);
```

Comment:
Personally, I would extract out the part of `removeFieldsWithIncompatibleTypes` that checks the types and re-use the extracted method here and in `removeFieldsWithIncompatibleTypes`. It would allow for more direct statement, like:
```
if (!fieldHasCompatibleTypes(field)) {
  throw ...
}
```

Reply:
Will do